From 3911d911c3530575832427cdada4f44296256fcc Mon Sep 17 00:00:00 2001
From: Chenming Huang <quic_chenhuan@quicinc.com>
Date: Sat, 11 Jan 2025 08:49:28 +0530
Subject: [PATCH] AP: Add support for 2-link MLO configuration

Add support to generate correct link configuration for AP MLD case.
1. Defines ap_mlo_links to store different links' channel/bandwidth.
2. Both links share the same interface.
3. The ctrl_interface directory is different. First config uses default
directory as legacy case. Another config use a "_1" suffix.
4. Do not force the same mac address for MLD mac and bssid.

Change-Id: I90ba767d9facd2235b2c9b756e4a48a7e8f54028
Signed-off-by: Chenming Huang <quic_chenhuan@quicinc.com>
---
 ap.c        | 281 +++++++++++++++++++++++++++++++++++++---------------
 sigma_dut.h |  17 ++++
 2 files changed, 219 insertions(+), 79 deletions(-)

diff --git a/ap.c b/ap.c
index 731b2cd..ecb3cbb 100644
--- a/ap.c
+++ b/ap.c
@@ -123,7 +123,7 @@ static void ath_ap_set_params(struct sigma_dut *dut);
 static int kill_process(struct sigma_dut *dut, char *proc_name,
 			unsigned char is_proc_instance_one, int sig);
 static int get_oper_centr_freq_seq_idx(struct sigma_dut *dut, int chwidth,
-				       int channel);
+				       int channel, int band);
 
 
 static int fwtest_cmd_wrapper(struct sigma_dut *dut, const char *arg,
@@ -964,7 +964,8 @@ static int get_bitmap_from_punct_chlist(struct sigma_dut *dut,
 		center_chan_idx = freq_to_chan(dut->ap_center_freq);
 	else
 		center_chan_idx = get_oper_centr_freq_seq_idx(dut, chwidth,
-							      dut->ap_channel);
+							      dut->ap_channel,
+							      dut->ap_band);
 	if (center_chan_idx < 0) {
 		sigma_dut_print(dut, DUT_MSG_ERROR,
 				"%s: couldn't get center channel index",
@@ -1026,6 +1027,7 @@ static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 	const char *ifname = get_main_ifname(dut);
 	char buf[128];
 	int subeamformermode = 0;
+	int mlo_config_band = -1;
 
 	/* Allow program to be overridden if specified in the ap_set_wireless
 	 * to support some 60 GHz test scripts where the program may be 60 GHz
@@ -1050,13 +1052,39 @@ static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 
 	val = get_param(cmd, "Interface");
 	if (val) {
-		if (strcasecmp(val, "5G") == 0)
+		if (strcasecmp(val, "5G") == 0) {
 			dut->ap_interface_5g = 1;
-		else
+			dut->ap_band = AP_BAND_5GHz;
+		} else if (strcasecmp(val, "24G") == 0) {
 			dut->ap_interface_2g = 1;
+			dut->ap_band = AP_BAND_24GHz;
+		} else if (strcasecmp(val, "6G") == 0) {
+			dut->ap_interface_6g = 1;
+			dut->ap_band = AP_BAND_6GHz;
+		}
 
-		if (dut->ap_interface_5g && dut->ap_interface_2g)
+		if (dut->ap_mode != AP_11be && dut->ap_interface_5g && dut->ap_interface_2g)
 			dut->ap_is_dual = 1;
+
+		if (dut->ap_mode == AP_11be) {
+			if (strcasecmp(val, "5G") == 0) {
+				dut->ap_mlo_links[AP_BAND_5GHz].configured = true;
+				dut->ap_interface_5g = 1;
+				mlo_config_band = AP_BAND_5GHz;
+			} else if (strcasecmp(val, "6G") == 0) {
+				dut->ap_mlo_links[AP_BAND_6GHz].configured = true;
+				dut->ap_interface_6g = 1;
+				mlo_config_band = AP_BAND_6GHz;
+			} else if (strcasecmp(val, "24G") == 0) {
+				dut->ap_mlo_links[AP_BAND_24GHz].configured = true;
+				dut->ap_interface_2g = 1;
+				mlo_config_band = AP_BAND_24GHz;
+			}
+			if ((dut->ap_interface_5g && dut->ap_interface_2g) ||
+			    (dut->ap_interface_6g && dut->ap_interface_2g) ||
+			    (dut->ap_interface_6g && dut->ap_interface_5g))
+				dut->ap_is_mlo = 1;
+		}
 	}
 
 	val = get_param(cmd, "CountryCode");
@@ -1123,15 +1151,19 @@ static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 		} else {
 			dut->ap_tag_channel[wlan_tag - 2] = atoi(val);
 		}
+
+		if (dut->ap_mode == AP_11be)
+			dut->ap_mlo_links[mlo_config_band].channel = atoi(val);
 	}
 
 	val = get_param(cmd, "ChnlFreq");
 	if (val) {
 		int freq = atoi(val);
 
-		if (is_6ghz_freq(freq))
+		if (is_6ghz_freq(freq)) {
 			dut->ap_band_6g = 1;
-		else
+			dut->ap_band = AP_BAND_6GHz;
+		} else
 			dut->ap_band_6g = 0;
 
 		dut->ap_center_freq = 0;
@@ -1241,7 +1273,7 @@ static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 		break;
 	}
 
-	if (dut->ap_is_dual)
+	if (dut->ap_is_dual || dut->ap_is_mlo)
 		dut->use_5g = 1;
 
 	val = get_param(cmd, "WME");
@@ -1518,25 +1550,48 @@ static enum sigma_cmd_result cmd_ap_set_wireless(struct sigma_dut *dut,
 
 	val = get_param(cmd, "WIDTH");
 	if (val) {
-		if (strcasecmp(val, "20") == 0)
-			dut->ap_chwidth = AP_20;
-		else if (strcasecmp(val, "40") == 0)
-			dut->ap_chwidth = AP_40;
-		else if (strcasecmp(val, "80") == 0)
-			dut->ap_chwidth = AP_80;
-		else if (strcasecmp(val, "160") == 0)
-			dut->ap_chwidth = AP_160;
-		else if (strcasecmp(val, "320") == 0)
-			dut->ap_chwidth = AP_320;
-		else if (strcasecmp(val, "80plus80") == 0) {
-			dut->ap_80plus80 = 1;
-			dut->ap_chwidth = AP_80_80;
-		} else if (strcasecmp(val, "Auto") == 0)
-			dut->ap_chwidth = AP_AUTO;
-		else {
-			send_resp(dut, conn, SIGMA_INVALID,
-				  "errorCode,Unsupported WIDTH");
-			return STATUS_SENT;
+		if (dut->ap_mode != AP_11be) {
+			if (strcasecmp(val, "20") == 0)
+				dut->ap_chwidth = AP_20;
+			else if (strcasecmp(val, "40") == 0)
+				dut->ap_chwidth = AP_40;
+			else if (strcasecmp(val, "80") == 0)
+				dut->ap_chwidth = AP_80;
+			else if (strcasecmp(val, "160") == 0)
+				dut->ap_chwidth = AP_160;
+			else if (strcasecmp(val, "320") == 0)
+				dut->ap_chwidth = AP_320;
+			else if (strcasecmp(val, "80plus80") == 0) {
+				dut->ap_80plus80 = 1;
+				dut->ap_chwidth = AP_80_80;
+			} else if (strcasecmp(val, "Auto") == 0)
+				dut->ap_chwidth = AP_AUTO;
+			else {
+				send_resp(dut, conn, SIGMA_INVALID,
+					  "errorCode,Unsupported WIDTH");
+				return STATUS_SENT;
+			}
+		} else {
+			if (strcasecmp(val, "20") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_20;
+			else if (strcasecmp(val, "40") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_40;
+			else if (strcasecmp(val, "80") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_80;
+			else if (strcasecmp(val, "160") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_160;
+			else if (strcasecmp(val, "320") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_320;
+			else if (strcasecmp(val, "80plus80") == 0) {
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_80_80;
+				dut->ap_80plus80 = 1;
+			} else if (strcasecmp(val, "Auto") == 0)
+				dut->ap_mlo_links[mlo_config_band].chwidth = AP_AUTO;
+			else {
+				send_resp(dut, conn, SIGMA_INVALID,
+					  "errorCode,Unsupported WIDTH");
+				return STATUS_SENT;
+			}
 		}
 	}
 
@@ -3243,6 +3298,8 @@ static enum sigma_cmd_result cmd_ap_set_security(struct sigma_dut *dut,
 	val = get_param(cmd, "BeaconProtection");
 	if (val)
 		dut->ap_beacon_prot = atoi(val);
+	else if (dut->ap_mode == AP_11be)
+		dut->ap_beacon_prot = 1;
 
 	val = get_param(cmd, "Transition_Disable");
 	if (val) {
@@ -8313,7 +8370,7 @@ static void set_ebtables_forward_drop(struct sigma_dut *dut,
 }
 
 
-static int check_channel(struct sigma_dut *dut, int channel)
+static int check_channel(struct sigma_dut *dut, int channel, int band)
 {
 	int channel_list[] = { 36, 40, 44, 48, 52, 60, 64, 100, 104, 108, 112,
 			       116, 120, 124, 128, 132, 140, 144, 149, 153, 157,
@@ -8327,7 +8384,7 @@ static int check_channel(struct sigma_dut *dut, int channel)
 	int num_chan, *chan_list;
 	int i;
 
-	if (dut->ap_band_6g) {
+	if (band == AP_BAND_6GHz) {
 		num_chan = ARRAY_SIZE(chan_list_6g);
 		chan_list = chan_list_6g;
 	} else {
@@ -8390,15 +8447,15 @@ static int get_6g_ch_op_class(int channel)
 
 
 static int get_oper_centr_freq_seq_idx(struct sigma_dut *dut, int chwidth,
-				       int channel)
+				       int channel, int band)
 {
 	int ch_base;
 	int period;
 
-	if (check_channel(dut, channel) < 0)
+	if (check_channel(dut, channel, band) < 0)
 		return -1;
 
-	if (dut->ap_band_6g)
+	if (band == AP_BAND_6GHz)
 		return get_oper_center_freq_6g(chwidth, channel);
 
 	if (channel >= 36 && channel <= 64)
@@ -8835,6 +8892,8 @@ enum sigma_cmd_result cmd_ap_config_commit(struct sigma_dut *dut,
 	enum driver_type drv;
 	const char *key_mgmt;
 	int conf_counter = 0;
+	int link_band = -1;
+	int i;
 	bool append_vht = false;
 	enum ap_mode mode;
 #ifdef ANDROID
@@ -8895,6 +8954,14 @@ write_conf:
 		if (run_system_wrapper(dut, "cp %s %s", f1, f2) != 0)
 			sigma_dut_print(dut, DUT_MSG_INFO,
 					"Failed to copy %s to %s", f1, f2);
+		if (dut->ap_mode == AP_11be) {
+			for (i = AP_BAND_MAX - 1; i >= 0; i--) {
+				if (dut->ap_mlo_links[i].configured) {
+					link_band = i;
+					break;
+				}
+			}
+		}
 
 #ifdef ANDROID
 		if (chmod(f2, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP) < 0)
@@ -8907,6 +8974,14 @@ write_conf:
 					"Error changing groupid");
 #endif /* ANDROID */
 	} else {
+		if (dut->ap_mode == AP_11be) {
+			for (i = 0; i < AP_BAND_MAX; i++) {
+				if (dut->ap_mlo_links[i].configured) {
+					link_band = i;
+					break;
+				}
+			}
+		}
 		ap_conf_path_1[0] = '\0';
 	}
 
@@ -8958,6 +9033,12 @@ write_conf:
 					fprintf(f, "hw_mode=a\n");
 				else
 					fprintf(f, "hw_mode=g\n");
+			} else if (dut->ap_is_mlo) {
+				if (link_band == AP_BAND_5GHz ||
+				    link_band == AP_BAND_6GHz)
+					fprintf(f, "hw_mode=a\n");
+				else
+					fprintf(f, "hw_mode=g\n");
 			} else {
 				fprintf(f, "hw_mode=a\n");
 			}
@@ -8981,6 +9062,9 @@ write_conf:
 	    (mode == AP_11ng || mode == AP_11na ||
 	     ((mode == AP_11ax || mode == AP_11be) && !dut->use_5g))) {
 		int ht40plus = 0, ht40minus = 0, tx_stbc = 0;
+		int chwidth = dut->ap_chwidth;
+		if (dut->ap_mode == AP_11be)
+			chwidth = dut->ap_mlo_links[link_band].chwidth;
 
 		fprintf(f, "ieee80211n=1\n");
 		if (mode == AP_11ax || mode == AP_11be)
@@ -8989,8 +9073,8 @@ write_conf:
 			fprintf(f, "ieee80211be=1\n");
 
 		if (mode == AP_11ng &&
-		    (dut->ap_chwidth == AP_40 ||
-		     (dut->ap_chwidth == AP_AUTO &&
+		    (chwidth == AP_40 ||
+		     (chwidth == AP_AUTO &&
 		      dut->default_11ng_ap_chwidth == AP_40))) {
 			if (dut->ap_channel >= 1 && dut->ap_channel <= 7)
 				ht40plus = 1;
@@ -9001,8 +9085,8 @@ write_conf:
 
 		/* configure ht_capab based on channel width */
 		if (mode == AP_11na &&
-		    (dut->ap_chwidth == AP_40 ||
-		     (dut->ap_chwidth == AP_AUTO &&
+		    (chwidth == AP_40 ||
+		     (chwidth == AP_AUTO &&
 		      dut->default_11na_ap_chwidth == AP_40))) {
 			if (is_ht40plus_chan(dut->ap_channel))
 				ht40plus = 1;
@@ -9014,11 +9098,11 @@ write_conf:
 			tx_stbc = 1;
 
 		/* Overwrite the ht_capab with offset value if configured */
-		if (dut->ap_chwidth == AP_40 &&
+		if (chwidth == AP_40 &&
 		    dut->ap_chwidth_offset == SEC_CH_40ABOVE) {
 			ht40plus = 1;
 			ht40minus = 0;
-		} else if (dut->ap_chwidth == AP_40 &&
+		} else if (chwidth == AP_40 &&
 			   dut->ap_chwidth_offset == SEC_CH_40BELOW) {
 			ht40minus = 1;
 			ht40plus = 0;
@@ -9035,6 +9119,12 @@ write_conf:
 	    (mode == AP_11ac ||
 	    ((mode == AP_11ax || mode == AP_11be) && dut->use_5g))) {
 		int ht40plus = 0, ht40minus = 0;
+		int chan = dut->ap_channel;
+		int chwidth = dut->ap_chwidth;
+		if (dut->ap_mode == AP_11be) {
+			chwidth = dut->ap_mlo_links[link_band].chwidth;
+			chan = dut->ap_mlo_links[link_band].channel;
+		}
 
 		fprintf(f, "ieee80211ac=1\n"
 			"ieee80211n=1\n");
@@ -9044,10 +9134,10 @@ write_conf:
 			fprintf(f, "ieee80211be=1\n");
 
 		/* configure ht_capab based on channel width */
-		if (dut->ap_chwidth != AP_20) {
-			if (is_ht40plus_chan(dut->ap_channel))
+		if (chwidth != AP_20) {
+			if (is_ht40plus_chan(chan))
 				ht40plus = 1;
-			else if (is_ht40minus_chan(dut->ap_channel))
+			else if (is_ht40minus_chan(chan))
 				ht40minus = 1;
 
 			fprintf(f, "ht_capab=%s%s\n",
@@ -9104,18 +9194,23 @@ write_conf:
 
 		fprintf(f, "mld_ap=1\n");
 		/* Use the same MLD and link MAC address */
-		fprintf(f, "mld_addr=%02x:%02x:%02x:%02x:%02x:%02x\n",
-			addr[0], addr[1], addr[2],
-			addr[3], addr[4], addr[5]);
-		fprintf(f, "bssid=%02x:%02x:%02x:%02x:%02x:%02x\n",
-			addr[0], addr[1], addr[2],
-			addr[3], addr[4], addr[5]);
+		if (!dut->ap_is_mlo) {
+			fprintf(f, "mld_addr=%02x:%02x:%02x:%02x:%02x:%02x\n",
+				addr[0], addr[1], addr[2],
+				addr[3], addr[4], addr[5]);
+			fprintf(f, "bssid=%02x:%02x:%02x:%02x:%02x:%02x\n",
+				addr[0], addr[1], addr[2],
+				addr[3], addr[4], addr[5]);
+		}
 	}
 
 	if (conf_counter == 1) {
 		char ifname2[50];
 
-		snprintf(ifname2, sizeof(ifname2), "%s_1", ifname);
+		if (mode != AP_11be)
+			snprintf(ifname2, sizeof(ifname2), "%s_1", ifname);
+		else
+			snprintf(ifname2, sizeof(ifname2), "%s", ifname);
 		if (if_nametoindex(ifname2) == 0) {
 			if (ap_create_or_remove_interface(dut, ifname2, true)
 			    >= 0) {
@@ -9134,19 +9229,27 @@ write_conf:
 	if (dut->bridge)
 		fprintf(f, "bridge=%s\n", dut->bridge);
 
-	if (dut->ap_is_dual && conf_counter == 1) {
+	if (dut->ap_mode == AP_11be) {
+		fprintf(f, "channel=%d\n", dut->ap_mlo_links[link_band].channel);
+	} else if (dut->ap_is_dual && conf_counter == 1) {
 		if (dut->ap_channel_1)
 			fprintf(f, "channel=%d\n", dut->ap_channel_1);
 		else
 			fprintf(f, "channel=%d\n", dut->ap_tag_channel[0]);
-	}
-	else
+	} else
 		fprintf(f, "channel=%d\n", dut->ap_channel);
 
-	if (sigma_hapd_ctrl)
-		fprintf(f, "ctrl_interface=%s\n", sigma_hapd_ctrl);
-	else
-		fprintf(f, "ctrl_interface=/var/run/hostapd\n");
+	if (sigma_hapd_ctrl) {
+		if (mode == AP_11be && conf_counter == 0 && (dut->ap_is_dual || dut->ap_is_mlo))
+			fprintf(f, "ctrl_interface=%s_1\n", sigma_hapd_ctrl);
+		else
+			fprintf(f, "ctrl_interface=%s\n", sigma_hapd_ctrl);
+	} else {
+		if (mode == AP_11be && conf_counter == 1)
+			fprintf(f, "ctrl_interface=/var/run/hostapd_1\n");
+		else
+			fprintf(f, "ctrl_interface=/var/run/hostapd\n");
+	}
 
 	if (dut->ap_ssid[0] && conf_counter == 0)
 		fprintf(f, "ssid=%s\n", dut->ap_ssid);
@@ -9747,13 +9850,27 @@ skip_key_mgmt:
 	      dut->program == PROGRAM_WPA3) &&
 	     dut->use_5g)) {
 		int vht_oper_centr_freq_idx;
+		int chan;
+		int chwidth;
+		int band;
+		if (dut->ap_mode == AP_11be) {
+			chan = dut->ap_mlo_links[link_band].channel;
+			chwidth = dut->ap_mlo_links[link_band].chwidth;
+			band = link_band;
+		} else {
+			chan = dut->ap_channel;
+			chwidth = dut->ap_chwidth;
+			band = dut->ap_band;
+		}
 
 		/* Do not try to enable VHT or higher channel bandwidth for HE
 		 * on the 2.4 GHz band when configuring a dual band AP that
 		 * does have VHT enabled on the 5 GHz radio. */
-		if (dut->use_5g) {
+		if (dut->ap_mode == AP_11be) {
+			if (band != AP_BAND_24GHz)
+				append_vht = true;
+		} else if (dut->use_5g) {
 			if (dut->ap_is_dual) {
-				int chan;
 
 				if (conf_counter)
 					chan = dut->ap_tag_channel[0];
@@ -9768,7 +9885,7 @@ skip_key_mgmt:
 		if (!append_vht)
 			goto skip_vht_parameters_set;
 
-		if (check_channel(dut, dut->ap_channel) < 0) {
+		if (band != AP_BAND_24GHz && check_channel(dut, chan, band) < 0) {
 			send_resp(dut, conn, SIGMA_INVALID,
 				  "errorCode,Invalid channel");
 			fclose(f);
@@ -9781,42 +9898,42 @@ skip_key_mgmt:
 			goto set_center_freq_idx_conf;
 		}
 
-		switch (dut->ap_chwidth) {
+		switch (chwidth) {
 		case AP_20:
 			dut->ap_vht_chwidth = AP_20_40_VHT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 20,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		case AP_40:
 			dut->ap_vht_chwidth = AP_20_40_VHT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 40,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		case AP_80:
 			dut->ap_vht_chwidth = AP_80_VHT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 80,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		case AP_160:
 			dut->ap_vht_chwidth = AP_160_VHT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 160,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		case AP_320:
 			dut->ap_vht_chwidth = AP_320_VHT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 320,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		default:
 			dut->ap_vht_chwidth = VHT_DEFAULT_OPER_CHWIDTH;
 			vht_oper_centr_freq_idx =
 				get_oper_centr_freq_seq_idx(dut, 80,
-							    dut->ap_channel);
+							    chan, band);
 			break;
 		}
 
@@ -9837,11 +9954,10 @@ set_center_freq_idx_conf:
 				vht_oper_centr_freq_idx);
 		}
 
-		if (dut->ap_band_6g)
+		if (band == AP_BAND_6GHz)
 			fprintf(f, "op_class=%d\n",
 				get_6g_ch_op_class(vht_oper_centr_freq_idx));
 
-
 		find_ap_ampdu_exp_and_max_mpdu_len(dut);
 
 		if (append_vht &&
@@ -9849,9 +9965,9 @@ set_center_freq_idx_conf:
 		     dut->ap_ldpc != VALUE_NOT_SET ||
 		     dut->ap_tx_stbc == VALUE_ENABLED || dut->ap_mu_txBF ||
 		     dut->ap_ampdu_exp || dut->ap_max_mpdu_len ||
-		     dut->ap_chwidth == AP_160 ||
-		     dut->ap_chwidth == AP_80_80 ||
-		     dut->ap_chwidth == AP_320)) {
+		     chwidth == AP_160 ||
+		     chwidth == AP_80_80 ||
+		     chwidth == AP_320)) {
 			fprintf(f, "vht_capab=%s%s%s%s%s%s",
 				dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
 				dut->ap_txBF ?
@@ -9861,9 +9977,9 @@ set_center_freq_idx_conf:
 				(dut->ap_tx_stbc == VALUE_ENABLED) ?
 				"[TX-STBC-2BY1]" : "",
 				dut->ap_mu_txBF ? "[MU-BEAMFORMER]" : "",
-				dut->ap_chwidth == AP_160 ||
-				dut->ap_chwidth == AP_320 ? "[VHT160]" :
-				(dut->ap_chwidth == AP_80_80 ?
+				chwidth == AP_160 ||
+				chwidth == AP_320 ? "[VHT160]" :
+				(chwidth == AP_80_80 ?
 				 "[VHT160-80PLUS80]" : ""));
 
 			if (dut->ap_ampdu_exp)
@@ -9983,7 +10099,7 @@ skip_vht_parameters_set:
 	}
 	fclose(f);
 
-	if (dut->ap_is_dual && conf_counter == 0) {
+	if ((dut->ap_is_dual || dut->ap_is_mlo) && conf_counter == 0) {
 		conf_counter++;
 		goto write_conf;
 	}
@@ -10086,8 +10202,8 @@ skip_vht_parameters_set:
 			 dut->use_hostapd_pid_file ?
 			 " -P " SIGMA_DUT_HOSTAPD_PID_FILE : "",
 			 dut->sigma_tmpdir,
-			 dut->ap_is_dual ? dut->sigma_tmpdir : "",
-			 dut->ap_is_dual ? "/sigma_dut-ap_0.conf" : "");
+			 (dut->ap_is_dual || dut->ap_is_mlo) ? dut->sigma_tmpdir : "",
+			 (dut->ap_is_dual || dut->ap_is_mlo) ? "/sigma_dut-ap_0.conf" : "");
 	}
 
 	sigma_dut_print(dut, DUT_MSG_DEBUG, "hostapd command: %s", buf);
@@ -10773,6 +10889,9 @@ static enum sigma_cmd_result cmd_ap_reset_default(struct sigma_dut *dut,
 		dut->ap_tag_channel[i] = 0;
 	}
 
+	for (i = 0; i < AP_BAND_MAX; i++)
+		dut->ap_mlo_links[i].configured = false;
+
 	drv = get_driver_type(dut);
 
 	program = get_param(cmd, "program");
@@ -10797,6 +10916,7 @@ static enum sigma_cmd_result cmd_ap_reset_default(struct sigma_dut *dut,
 	dut->ap_addba_reject = VALUE_NOT_SET;
 	dut->ap_noack = VALUE_NOT_SET;
 	dut->ap_is_dual = 0;
+	dut->ap_is_mlo = 0;
 	dut->ap_mode = AP_inval;
 	dut->ap_mode_1 = AP_inval;
 	dut->ap_channel_1 = 0;
@@ -10815,6 +10935,7 @@ static enum sigma_cmd_result cmd_ap_reset_default(struct sigma_dut *dut,
 	dut->ap_ampdu_exp = 0;
 	dut->ap_max_mpdu_len = 0;
 	dut->ap_band_6g = 0;
+	dut->ap_band = -1;
 	dut->ap_punct_bitmap = 0;
 	dut->ap_center_freq = 0;
 
@@ -10831,6 +10952,7 @@ static enum sigma_cmd_result cmd_ap_reset_default(struct sigma_dut *dut,
 	dut->ft_bss_mac_cnt = 0;
 	dut->ap_interface_5g = 0;
 	dut->ap_interface_2g = 0;
+	dut->ap_interface_6g = 0;
 	dut->ap_pmf = AP_PMF_DISABLED;
 
 	dut->wsc_fragment = 0;
@@ -14910,7 +15032,8 @@ static int hostapd_chan_switch(struct sigma_dut *dut, const char *ifname,
 								    is_6g);
 			center_freq_idx = get_oper_centr_freq_seq_idx(dut,
 								      chwidth,
-								      channel);
+								      channel,
+								      dut->ap_band);
 		} else {
 			sec_channel_offset = 1;
 			center_freq_idx = channel + 10;
@@ -15307,7 +15430,7 @@ static int mac80211_vht_chnum_band(struct sigma_dut *dut, const char *ifname,
 	if (result)
 		chwidth = atoi(result);
 
-	center_freq_idx = get_oper_centr_freq_seq_idx(dut, chwidth, channel);
+	center_freq_idx = get_oper_centr_freq_seq_idx(dut, chwidth, channel, dut->ap_band);
 	if (center_freq_idx < 0) {
 		free(token);
 		return -1;
@@ -15357,7 +15480,7 @@ mac80211_he_tx_bandwidth(struct sigma_dut *dut, struct sigma_conn *conn,
 
 	width = atoi(val);
 	center_freq_idx = get_oper_centr_freq_seq_idx(dut, width,
-						      dut->ap_channel);
+						      dut->ap_channel, dut->ap_band);
 	if (center_freq_idx < 0)
 		return ERROR_SEND_STATUS;
 
diff --git a/sigma_dut.h b/sigma_dut.h
index c32f0c9..2a8f4a0 100644
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -538,6 +538,19 @@ struct mdnss_discovery_info {
 };
 #endif /* ANDROID_MDNS */
 
+struct ap_mlo_link {
+	bool configured;
+	int chwidth;
+	int channel;
+};
+
+enum ap_band {
+	AP_BAND_24GHz,
+	AP_BAND_5GHz,
+	AP_BAND_6GHz,
+	AP_BAND_MAX,
+};
+
 enum sigma_akm_suites {
 	SIGMA_AKM_WPA_PSK = 0,
 	SIGMA_AKM_FT_PSK,
@@ -609,10 +622,12 @@ struct sigma_dut {
 	char *p2p_ifname_buf;
 	int use_5g;
 	int ap_band_6g;
+	int ap_band;
 	int ap_center_freq;
 	int ap_punct_bitmap;
 	int sta_2g_started;
 	int sta_5g_started;
+	struct ap_mlo_link ap_mlo_links[AP_BAND_MAX];
 
 	int s; /* server TCP socket */
 	int debug_level;
@@ -938,6 +953,7 @@ struct sigma_dut {
 	int ap_name;
 	int ap_interface_5g;
 	int ap_interface_2g;
+	int ap_interface_6g;
 	int ap_assoc_delay;
 	int ap_btmreq_bss_term_tsf;
 	int ap_fils_dscv_int;
@@ -1057,6 +1073,7 @@ struct sigma_dut {
 	const char *set_macaddr;
 	int tmp_mac_addr;
 	int ap_is_dual;
+	int ap_is_mlo;
 	enum ap_mode ap_mode_1;
 	enum ap_chwidth ap_chwidth_1;
 	int ap_channel_1;
-- 
2.17.1

